src/compute-client/src/plan/interpret/api.rs

```diff
@@ -0,0 +1,650 @@
+        input: Self::Domain,
+        top_k_plan: &TopKPlan,
+    ) -> Self::Domain;
+
+    fn negate(&self, ctx: &Context<Self::Domain>, input: Self::Domain) -> Self::Domain;
+
+    fn threshold(
+        &self,
+        ctx: &Context<Self::Domain>,
+        input: Self::Domain,
+        threshold_plan: &ThresholdPlan,
+    ) -> Self::Domain;
+
+    fn union(&self, ctx: &Context<Self::Domain>, inputs: Vec<Self::Domain>) -> Self::Domain;
+
+    fn arrange_by(
+        &self,
+        ctx: &Context<Self::Domain>,
+        input: Self::Domain,
+        forms: &AvailableCollections,
+        input_key: &Option<Vec<MirScalarExpr>>,
+        input_mfp: &MapFilterProject,
+    ) -> Self::Domain;
+}
+
+/// An [Interpreter] context.
+pub type Context<Domain> = BTreeMap<LocalId, ContextEntry<Domain>>;
+
+/// An entry in an [Interpreter] context.
+///
+/// Each entry corresponds to a binding identified by a [LocalId] that is
+/// visible in the current context.
+#[derive(Debug)]
+pub struct ContextEntry<Domain> {
+    /// Is this entry correspond to a recursive binding or not.
+    pub is_rec: bool,
+    /// The domain value associated with this binding.
+    pub value: Domain,
+}
+
+impl<Domain> ContextEntry<Domain> {
+    fn of_let(value: Domain) -> Self {
+        Self {
+            is_rec: false,
+            value,
+        }
+    }
+
+    fn of_let_rec(value: Domain) -> Self {
+        Self {
+            is_rec: true,
+            value,
+        }
+    }
+}
+
+/// A lattice with existing `top` and `bottom` elements.
+pub trait BoundedLattice: Lattice {
+    /// The top element of this lattice (represents the most general
+    /// approximation).
+    fn top() -> Self;
+
+    /// The bottom element of this lattice (represents the most constrained
+    /// approximation).
+    fn bottom() -> Self;
+}
+
+/// The maximum of iterations of running lattice-based dataflow inference for
+/// the LetRec nodes before falling back to a conservative estimate of all
+/// bottom() elements for all recursive bindings.
+const MAX_LET_REC_ITERATIONS: u64 = 100;
+
+/// A wrapper for a recursive fold invocation over a [Plan] that cannot
+/// mutate its input.
+#[allow(missing_debug_implementations)]
+pub struct Fold<I, T>
+where
+    I: Interpreter<T>,
+{
+    interpret: I,
+    ctx: Context<I::Domain>,
+}
+
+impl<I, T> Fold<I, T>
+where
+    I: Interpreter<T>,
+    I::Domain: BoundedLattice + Clone,
+{
+    pub fn new(interpreter: I) -> Self {
+        Self {
+            interpret: interpreter,
+            ctx: Context::default(),
+        }
+    }
+
+    /// An immutable fold (structural recursion) over a [Plan] instance.
+    ///
+    /// Runs an abstract interpreter over the given `expr` in a bottom-up
+    /// manner, keeping the `ctx` field of the enclosing field up to date, and
+    /// returns the final result for the entire `expr`.
+    pub fn apply(&mut self, expr: &Plan<T>) -> I::Domain {
+        use Plan::*;
+        match expr {
```

Comment:
Should this use guarded recursion?

Reply:
Yes, I've listed this as an open question - the issue is that if I use guided recursion the return type needs to change, and in principle I think we have an option here to use stack-based recursion.

Reply:
I added guided recursion to the two modules in separate commits for easier reviewing.